lib/vagrant/util/file_mutex.rb

```diff
@@ -0,0 +1,29 @@
+module Vagrant
+  module Util
+    class FileMutex
+      def initialize(mutex_path)
+        @mutex_path = mutex_path
+      end
+
+      def with_lock(&block)
+        lock
+        block.call
+      ensure
+        unlock
+      end
+
+      def lock
+        if File.file?(@mutex_path)
```

Comment:
There ends up being a race condition here between checking for the existence of the file, and creating the file. Instead, it would be better to open the file and then [lock it](https://ruby-doc.org/3.1.3/File.html#method-i-flock) to ensure an exclusive lock.